Add System.Diagnostics.DiagnosticSource netstandard guidance (#43269)
dotnet/docs: docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md

Applications that target .NET 5 and later already have the necessary distributed tracing APIs included. For apps targeting older
.NET versions, add the [System.Diagnostics.DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/)

> version 5 or greater. For libraries targetting netstandard, we recommend referencing the latest version of NuGet package.

Author: ```suggestion
version 5 or greater. For libraries targetting netstandard, we recommend referencing the latest version of the NuGet package.
```

> version 5 or greater. For libraries targetting netstandard, we recommend referencing the latest version of the NuGet package.

Reviewer: ```suggestion
version 5 or greater. For libraries targeting netstandard, we recommend referencing the latest version of the NuGet package.
```

> version 5 or greater. For libraries targetting netstandard, we recommend referencing the latest version of the NuGet package.

Reviewer: Is the latest version always the right one? If you have customers on `net8.0` and your library references the `9.0.0` version, it will "lift" the assembly out of the shared framework.

Reviewer 2: Upgrade will be required only if the library need a feature in recent version that is not exist in the current targeting version.

Reviewer: If the library references the 9.0 version, System.Diagnostics.DiagnosticSource will be lifted out of the shared framework whether or not the library uses a new feature or not.

Maybe we should say to reference the oldest LTS version that is currently supported? For example, this month that would be `net6.0`. But next month that would be `net8.0`. (since .NET 6 is going out of support next month).

Reviewer 2: Are we still talking about `nestandard` only case? or you are talking generally? 

- When targeting .NET, use the version of `System.Diagnostics.DiagnosticSource` that ships with the framework, unless you need a feature not supported in that version. In such cases, reference the earliest supported version that includes the required feature.
- When targeting `netstandard`, reference the earliest supported version of `System.Diagnostics.DiagnosticSource` that includes the feature you need.

"Supported version of `System.Diagnostics.DiagnosticSource`" refers to the version that aligns with a supported .NET version.

> version 5 or greater. For libraries targetting netstandard, we recommend referencing the latest version of the NuGet package.

Author: ```suggestion
version 5 or greater. For libraries targetting netstandard, we recommend referencing the oldest LTS version of the package which is still supported and contains the APIs your library needs.
```
Does that work? If not feel free to use the suggestion feature and propose the exact text you want.

Reviewer 2: I am afraid this does not cover the case if the required feature is in non-LTS. Implicitly telling don't use it. My comment https://github.com/dotnet/docs/pull/43269#discussion_r1825175898 may cover all cases? I am not suggesting taking it as it is. Just cover the cases we need to talk about.

Author: Edited to remove the LTS, how about now?

> 

Reviewer 2: I would like to add some disclaimer sentence telling, 

Here’s a clearer version:

If your library targets `netstandard` and references a recent version of `System.Diagnostics.DiagnosticSource`, users consuming the `netstandard` package may see a warning if they’re targeting an unsupported .NET version. Users may need to upgrade to a supported .NET version to avoid issues.

Author: Do you still think this is needed now that the guidance is use the oldest? The primary purpose of the doc is teaching people how to use the instrumentation APIs so it feels like this disclaimer is getting somewhat far from the main topic.

Reviewer 2: I am fine not adding it for now.

> version 5 or greater. For libraries targetting netstandard, we recommend referencing the oldest version of the package which is still supported and contains the APIs your library needs.

Author: ```suggestion
version 5 or greater. For libraries targeting netstandard, we recommend referencing the oldest version of the package which is still supported and contains the APIs your library needs.
```
